core/src/main/java/org/apache/carbondata/core/datamap/DistributableDataMapFormat.java

```diff
@@ -85,33 +143,67 @@
       private ExtendedBlocklet currBlocklet;
       private List<DataMap> dataMaps;
 
-      @Override public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+      @Override
+      public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
           throws IOException, InterruptedException {
-        distributable = (DataMapDistributableWrapper) inputSplit;
-        // clear the segmentMap and from cache in executor when there are invalid segments
-        if (invalidSegments.size() > 0) {
-          DataMapStoreManager.getInstance().clearInvalidSegments(table, invalidSegments);
-        }
-        TableDataMap tableDataMap = DataMapStoreManager.getInstance()
-            .getDataMap(table, distributable.getDistributable().getDataMapSchema());
+        DataMapDistributableWrapper distributable = (DataMapDistributableWrapper) inputSplit;
+        distributable.getDistributable().getSegment().setCacheable(!isFallbackJob);
+        distributable.getDistributable().getSegment().setReadCommittedScope(readCommittedScope);
+        List<Segment> segmentsToLoad = new ArrayList<>();
+        segmentsToLoad.add(distributable.getDistributable().getSegment());
         if (isJobToClearDataMaps) {
-          // if job is to clear datamaps just clear datamaps from cache and return
-          DataMapStoreManager.getInstance()
-              .clearDataMaps(table.getCarbonTableIdentifier().getTableUniqueName());
-          // clear the segment properties cache from executor
-          SegmentPropertiesAndSchemaHolder.getInstance()
-              .invalidate(table.getAbsoluteTableIdentifier());
-          blockletIterator = Collections.emptyIterator();
+          if (StringUtils.isNotEmpty(dataMapToClear)) {
+            List<TableDataMap> dataMaps =
+                DataMapStoreManager.getInstance().getAllVisibleDataMap(table);
+            List<TableDataMap> dataMapsTobeRemoved = new ArrayList<>();
+            for (TableDataMap tableDataMap : dataMaps) {
+              if (tableDataMap != null && dataMapToClear
+                  .equalsIgnoreCase(tableDataMap.getDataMapSchema().getDataMapName())) {
+                tableDataMap.deleteSegmentDatamapData(
+                    ((DataMapDistributableWrapper) inputSplit).getDistributable().getSegment()
+                        .getSegmentNo());
+                dataMapsTobeRemoved.add(tableDataMap);
+              }
+            }
+            dataMaps.removeAll(dataMapsTobeRemoved);
```

Comment:
This remove will not work as tableDatamap does not override equals properly. Please handle it.

Reply:
removed based on index